File: Python/23-二叉搜索树的后序遍历序列.py
# -*- coding:utf-8 -*-
class Solution:
    def VerifySquenceOfBST(self, sequence):
        # write code here
        if not sequence:
            return False
        def helper(s, low, high):
            if low >= high:
                return True
            pivot = high - 1
            while pivot >= low and s[pivot] > s[high]:
                pivot -= 1
            for i in range(low, pivot + 1):
                if s[i] > s[high]:
                    return False
            return helper(s, low, pivot) and helper(s, pivot + 1, high - 1)
        return helper(sequence, 0, len(sequence) - 1)

File: C++/23-二叉搜索树的后序遍历序列.cpp
class Solution {
public:
    bool VerifySquenceOfBST(vector<int> sequence) {
        if(sequence.empty())
            return false;
        return helper(sequence, 0, sequence.size() - 1);
    }
    
    bool helper(vector<int> s, int low, int high){
        if(low >= high)
            return true;
        int pivot = high - 1;
        while(low <= pivot && s[pivot] > s[high])
            --pivot;
        for(int i = low; i <= pivot; i++)
            if(s[i] > s[high])
                return false;
        return helper(s, low, pivot) && helper(s, pivot + 1, high - 1);
    }
};